Tutorial copy-editing (#1)
jkshenton/castep-docs: docs/tutorials/NEB/neb_tutorial.md

## Setting up the structures

### Setting up the initial and final states
These are the local minima structures you want to find the barrier between. They can be equivalent structures, e.g. differing by a symmetry operation, such reflection through a mirror plane, or they can be configurations with different energies (in which case your forward and reverse barriers will not be the same!). Two crucial conditions must be met **before running the NEB calculation**:

> 1. The end point structures *[and any intermediates?]* must be fully relaxed (geometry optimised) *[using the same parameters as NEB presumably]* before you start the NEB.

Reviewer: You typically generate the intermediate (guess) structure based the (optimised) end points (e.g. through linear interpolation). The intermediate structure will be relaxed during the course of the NEB, so it doesn't need to be optimised beforehand. In fact, if you tried to optimise it independently, it might well roll down to one of the end point structures. 

Yes, the XC functional, basis set, k-points etc should all be the same between the initial end point relaxations and the NEB. But of course some things in the .param file have to differ (e.g. the TASK).

Author: Yes, I realised afterwards that you can't optimise the intermediate! I think it's useful to state this explicitly in a tutorial context. I would also state the point about consistent calculations. It's obvious to an expert, but again, you can be explicit in a tutorial.

Reviewer: ```suggestion
1. The end point structures must be fully relaxed (geometry optimised) before you start the NEB. Note that both the initial relaxations and the actual NEB calculation must use consistent parameters (e.g. basis set, pseudopotentials, k-points, XC functional etc.).
```

### Setting up the transition state guess
CASTEP will linearly interpolate between the initial and final structures provided to generate the first guess of the minimum energy pathway.

> This works well in many cases. However, if your minimum energy pathway contains e.g. a rotation of a (group of) atom(s) around a bond or something similar, then the linear path between initial and final configurations may be a bad starting guess. *[Not clear to me what a bad starting guess means - presumably if the interpolated intermediate is too high in energy, things will go wrong? Obviously a C2 rotation can't be interpolated, but a small rotation, perhaps C4 might be fine?]* Intermediate structures can also be used to bias your results to one pathway over another.

Reviewer: Yes, your example of the C2 vs C4 rotations is a good one.

Reviewer: ```suggestion
This works well in many cases. However, if your minimum energy pathway contains e.g. a rotation of a (group of) atom(s) around a bond or something similar, then the linear path between initial and final configurations may be a bad starting guess/non-physical structure. For example, if your end points correspond to a C2 rotation, then a linearly interpolated intermediate structure would not be a physical one. It's important to visualise or otherwise check all of your structures before running a calculation.

Intermediate structures can also be chosen in such a way as to bias your results to one pathway over another.
```

### How many images do I need?

> It depends... It's usually easier to start with a small number of images and increase if you need to. *[A bit vague.]*

Reviewer: Yes, it's hard to come up to general answers to these kinds of questions...

What I was going for here was that: if you start with loads of images then you might never get close to the converged TS because the optimisation takes too long. If you instead start with a small number and get a reasonable TS structure, you can then use that as your initial intermediate structure in a restarted calculation with more images. Starting small can also quickly give you can rough sense of the energy landscape and that will help determine the actual number of images needed.

Author: I would just say this. A tutorial gives you the space to explain the thinking behind doing the calculations well.

Reviewer: ```suggestion
It depends... It's usually preferable to start with a small number of images and increase the number if you need to. Starting small can quickly give you a rough sense of the energy landscape and that will determine the actual number of images needed. 

In addition, if you start with too many images then you might never get close to the converged TS because the optimisation takes too long. If you instead start with a small number and get a reasonable TS structure, you can then use that as your initial intermediate structure in a restarted calculation with more images.
```

> Using many (i.e. > 15) will result in slow convergence, but may lead to a more accurate minimum energy pathway. *[Presumably more images will never give a worse result? Would it be better expressed in terms of a rule of thumb e.g. 10 x the number of expected transition states?]*

Reviewer: I think it's quite tricky to give a rule of thumb for this, at least I don't feel qualified to do so.

Author: Fair enough - I think this addressed by the point above.

Reviewer: ```suggestion
Using many (i.e. > 15) will result in slow convergence, but may lead to a more accurate minimum energy pathway for complex barriers. If you do find a particularly complex pathway, however, you might be better off splitting the work up into more than one NEB calculation, each with fewer images. Just remember to always fully relax the end points (even if the end point is a meta-stable local minimum)!  
```

> Using too few (i.e. < 5) might fail to find the minimum energy pathway. *[Not clear to me what this means. Partly because I don't have a clear idea of what the "spacing" between images means.]*

Reviewer: Maybe this recommendation should be changed anyway -- in principle even one image can be sufficient in the case of a very simple path. In the case of a complex pathway, though (e.g. >1 TS), you'll need more.

Author: Yes, this was confusing me. It wasn't obvious how to think about the number of images. Can it be viewed in terms of "interpolated" between fixed points start / intermediate / end? The main documentation says it is important to actively choose N (rather than accept the default, 20?), but I didn't know what to with this instruction. You could explain how you chose N for ammonia (easier in terms of a concrete example).

Reviewer: OK I'm going to add a suggested change including a note to myself to elaborate and add something else: should you use an even or odd number of images. (So that I can pull in all these changes and then make more substantial changes on my fork directly.

```suggestion
Using too few (i.e. < 5) might fail to find the minimum energy pathway if the energy landscape is complex.

TODO: add guidance on odd vs even number of images (touching on how the spacing is calculated) and on how many you need for this simple example -- maybe an exercise for the user? You can actually get the same barrier height with just one image in this simple case! 
```

%ENDBLOCK IONIC_CONSTRAINTS
```

> to the `nh3.cell` file.  *[Isn't there a more compact way now to specify a fixed atom?]*

Reviewer: @jryates ?

Reviewer 2: putting
 fix N
in the block is the same as those three lines. See https://www.tcm.phy.cam.ac.uk/~jry20/site/documentation/Geometry_Optimisation/constraints/

Reviewer: Thanks - Let's change it to `fix N` and reference the constraints page.

Author: Yes, I think it's important that the tutorials are consistently reinforcing good practice / exploiting new capability. 

@jryates Can we assume that everybody will be using CASTEP versions with this functionality? It would be good to include notes saying when features were introduced into CASTEP versions (something the Python documentation does well).

Reviewer 2: This was in the codebase about 8-9 years ago - so we are fine. I take the general point.

 
in the `.castep` output file.

*[Shouldn't this be referenced in the main documentation, as it looks like a specific feature of the NEB code?]*

### Do I need climbing NEB?

> The climbing NEB method modifies the force on the highest-energy image (usually close to the middle of the band) *[? What's a band?]* such that it tries to climb *up* the barrier. 

Reviewer: The B in NEB :) 

Could remove everything in parentheses to remove confusion. 
```suggestion
The climbing NEB method modifies the force on the highest-energy image such that it tries to climb *up* the barrier. 
```

> This is very useful to get accurate barrier estimates, but care must be taken when more complex pathways (i.e. with multiple maxima) are found.  *[Don't get this!]*

Reviewer: I could be wrong, but if you have multiple TS states, the CI NEB will still only make one of the images climb up, so you might get a distorted path as a result. In such cases it's probably best to break the path into multiple single-TS NEB calculations, fully relaxing any meta-stable intermediate structures.

Author: OK, should have connected "band" to NEB, but it's not clear to me what this band *is*. On the one hand, I visualised the band as the forces that help find the solution, on the other is it the solution itself - the final energy profile that "hugs" the minimum energy path.

Intuitively it does seem reasonable that a complex profile might need to be split into individual steps between metastable transition states.

Presumably things also get tricky if there are multiple pathways? You can presumably have a global vs. local minimisation problem? 

I think this is getting beyond the scope of the tutorial. I would emphasise that the tutorial is geared towards a simple case where you are not expecting metastable intermediates (a separate tutorial if demand?).

I would clarify what is meant by "climbing NEB method (or mode?)" vs. it's opposite.

Reviewer: I agree that it goes beyond the scope of the tutorial. I'll add a TODO item below to clarify this part of the tutorial and include advice about starting with regular NEB and turning on climbing image after a few steps.

In our example the path is quite straightforward and it makes sense to use the climbing image method. We therefore set:

`TSSEARCH_NEB_CLIMBING: TRUE`

> in the .param file. *[This is the default surely? Perhaps clearer in main documentation giving an example of when you *would* set False]* 

Reviewer: `TSSEARCH_NEB_CLIMBING` is set to False by default, according to the keywords table in the main docs.

Author: It seems odd if the default value needs to be changed for a simple example - the CASTEP philosophy is to Do the Right Thing out of the box. Either the default would be better changed, or the reasons for changing it from the default should be clear. At the moment, I don't really understanding what "climbing mode" means. What exactly are you climbing, a potential surface?

Reviewer: I think defaulting to climbing mode off is sensible since it prevents some possible convergence issues. Since the image chosen to be 'climbing' is whichever happens to the highest energy at a given step, in some cases, especially during the early stages of the optimisation, the climbing image might not always stay the same. In the worst cases this can lead to a sort of oscillating behaviour and convergence will be difficult. 

It's generally safer to start with climbing mode off, get something that is converged (or almost), then to switch on climbing mode and restart. Perhaps we should state that somewhere.

See the discussions on these pages for more: 

- https://theory.cm.utexas.edu/henkelman/research/saddle/neb/
- https://theory.cm.utexas.edu/vtsttools/neb.html#climbing-image

Reviewer: ```suggestion
in the .param file.

TODO: clarify what happens when you set climbing to true.  

TODO: include advice/explanation about starting with regular NEB and turning on climbing image after a few steps. 
```

Author: Yes, but remove comment in L139.

## Running the NEB

### Parallelisation
Because the NEB calculation essentially involves *N* independent CASTEP calculations (where *N* is the number of images), we can make very efficient use of high-performance computers. This is done via "task farming", running the calculations for the *N* images independently. For example, if one CASTEP calculation fits on one node, and you have 7 images, then you could run on a total of 7 nodes by setting the keyword:

`NUM_FARMS: 7`

in the .param file. 

> In general, setting `NUM_FARMS` to the number of NEB images will lead to good performance, provided the available computing resources can be split neatly by this number.  *[Don't follow the last phrase. Surely all you need are 7 nodes?]*

Reviewer: I guess you might run into performance issues if you cannot request as many nodes as you have images. For example if you have just one node available, then setting NUM_FARMS to 7 would split the workload 7 ways on that node, which might well be inefficient. (?)

Author: OK. I think this needs reframing. If you don't set NUM_FARMS, are the calculations does sequentially, without trying to parallelise? 

In this case, the question really comes down to whether your HPC set-up is sufficiently large for task farming to be useful. The no brainer is N nodes are available. If you had a small calculation and large nodes (e.g. 48 processors), then you might set NUM_FARMS to 4 say? 

But this discussion might be best in the more general documentation. I think task farming is used for J calculations? @jryates

Reviewer: ```suggestion
A good rule of thumb is to aim for `NUM_FARMS` = number of NEB images. That way each 'farm' is responsible for one NEB image. Since each farm will effectively run a series of single-point calculations on each NEB image they are responsible for, the compute resources of that farm should be appropriately sized for such calculations. Too small and you could get an out-of-memory error, too large and you might end up over-parallelising, leading to reduced overall performance. i.e. you should aim to run on a total of `NUM_FARMS` times as many cores as you would use for a single-point calculation on one of the structures.  

This also assumes that your available compute resources can be neatly split into `NUM_FARMS`. The easy case is when each farm can run on a single node; i.e. `NUM_FARMS` = number of NEB images = number of nodes. Note that the total number of cores available must be divisible by `NUM_FARMS`. 

If you choose `NUM_FARMS` < number of images, each of the images will run sequentially on one farm as soon as that farm becomes available. If the number of images is not a multiple of `NUM_FARMS`, some farms will end up being responsible for more images than others, leading poor overall efficiency.

If you choose `NUM_FARMS` > number of images, then some farms will be left with nothing to do, resulting in poor overall efficiency.

Setting `NUM_FARMS = 1` is equivalent to not setting `NUM_FARMS`.
```

Author: L154: "However, if you only have 2 nodes available, you cannot choose `NUM_FARMS = 7` since your compute resources cannot be distributed evenly across the farms."
What if your nodes have 14 processors? I would only use "cannot" or "must not" for really Bad Things, "cannot" = this will give an unrecoverable error. "must not" = this is really bad. You might get an invalid answer. "should not" = this is really not recommended. On your head be it. Something like "it wouldn't make sense to" for something that was a bit silly but not jeopardising your calculation.

L156: "poor performance" -> "reduced overall performance" I would use "poor performance" to mean "this is a really dumb thing to do since your calculation will run a factor of N times more slowly". If you have 3 farms and 8 images, it's not going to be that bad surely? 

"the N images will be distributed amongst the farms as evenly as possible". Presumably the images are run sequentially on each farm, rather than competing?

Is not setting NUM_FARMS equivalent to NUM_FARMS = 1?

L158: wouldn't the "excess" farms have nothing to do (rather than "being idle for a lot of time")?  

It is probably worth highlighting that if you set NUM_FARMS = N then you probably want to request N x as much processor resource, i.e. if you would run a single-point calculation on 1 compute node, then you would probably want N nodes?

Reviewer: I've updated my suggested changes -- hopefully that clears things up?

### Is the calculation converging?
You can monitor the convergence by searching for "Max NEB force" within the .castep output file.

Common convergence issues include:

- Not having consistent order in the initial and final structures, i.e. non-physical initial path
- Too many (or too few) images  *[As above - not clear what this means, so looks odd.]*
- Initial or final state (endpoints) not being fully relaxed structures

### Restarting / Continuing 

Sometimes the calculation will run out of time/number of iterations before the NEB calculation is finished. 
You will probably see a "Failed to converge" message in the .castep output in that case; always read through your output files!
*[Not obvious in .ts file? Shouldn't need to positively check for unconverged results.]*

If you re-run the calculation without changing anything, it will simply start from the beginning again -- probably not what you want!

> In order restart the NEB from the last checkpointed state, you need to **explicitly** set the name of the .check file you want to continue from. *[In .param?]* In our example that would be:

Reviewer: ```suggestion
In order restart the NEB from the last checkpointed state, you need to **explicitly** set the name of the .check file you want to continue from in your .param file. In our example that would be:
```

```
continuation: nh3.check
```

In the .castep output you would then see: `Coordinates of NEB images loaded from checkpoint file`. 

!!! warning
      Unlike other CASTEP tasks, setting `continuation: default` does not seem to work for continuing/restarting NEB calculations! You must explicitly set the name of checkpoint file.

## Analysis

Once you have a converged NEB, the provided Python utility `readTS` can be used to parse and analyse the NEB. 

> To add the readTS module to your PYTHONPATH, you can find a setup.py script in the `castep/Utilities/readts` directory of your CASTEP source directory. Running `python setup.py` should install the module to your path. You can also manually add it by: *[Add to what? .bashrc or current (bash) console?]* 

Reviewer: ```suggestion
To add the readTS module to your PYTHONPATH, you can find a setup.py script in the `castep/Utilities/readts` directory of your CASTEP source directory. Running `python setup.py` should install the module to your path. You can also manually add it by inserting the following line into your ~/.bashrc or equivalent and restarting the terminal: 
```

Author: Yes, but "restart" -> "restarting"

``` bash
export PYTHONPATH="/path/to/castep/Utilities/readts:$PYTHONPATH"
```

(changing the `/path/to/castep/` bit to wherever CASTEP is on your machine).

> You can then extract the NEB path using something like this python code:  *[This looks sufficiently general that you could supply the analysis.py script?]*

Reviewer: It's provided in the tar file with the tutorial files and the full script is shown here -- do we also need a separate link to download .py file?

Author: OK. But then there's not much point in giving the source of analysis.py in the tutorial. I would rename analysis to analysisTS or something more meaningful, and refer to reader to the script in the tutorial bundle. it would be more useful in the tutorial to highlight anything that the user might want to tweak / customise.

 
in the `.castep` output file.

> *[Shouldn't this be referenced in the main documentation, as it looks like a specific feature of the NEB code?]*

Reviewer: ```suggestion
```

Sure, I can add something about it there.

![NEB Plot](../../img/neb.png)

> *[Do the green lines have any significance?]*

Reviewer: ASE NEB plots show these green tangent lines -- they can be useful to see if something is going weird between two images. For this simple path they don't really add much, but it's more work to get rid of them than to keep them.

Author: Again, I would just give the 1st sentence.  You just don't want "unexplained stuff" in a tutorial.

Reviewer: ```suggestion
The green lines show the tangents at those points; these can be useful indicators of convergence. 
```

The barrier we obtained is about 0.23 eV, which compares well with the value of 0.25 eV (24.2 kJ/mol) [from Wikipedia](https://en.wikipedia.org/wiki/Pyramidal_inversion#Energy_barrier).

We might also notice the slight difference in energy between the two end-points of the path (about 1 meV). These should be identical, so any difference can be used as a very rough guide for the amount of numerical noise in the geometry optimisations.

> *[Surely this is more about the initial calculations rather than the NEB? Is this noise figure useful?]*

Reviewer: I'd say that the noise in the end points (from the initial geom optimisations as you say), gives you a rough handle how converged the calculations are. So the noise on any subsequent geometry optimisations (including the subsequent NEB calculation in this case)  should be a similar magnitude, unless you change your parameters (XC, basis set, pseudo, force tolerance etc.) of course. 

If equivalent structures (the end points in this example) were very different in energy, it would suggest you need to tighten up your convergence criteria or force/energy tolerances etc.

Author: Again, I think your expanded explanation helps the reader under the significance of the point you are making.

Reviewer: ```suggestion
If equivalent structures (the end-points in this example) were found to have very different energies, you would need to go back and tighten up your convergence criteria or force/energy tolerances in the initial geometry optimisations and re-run the NEB calculation with these tightened parameters. 
```

![NEB Plot](../../img/neb.png)

The green lines show the tangents at those points; these can be useful indicators of convergence. 

The barrier we obtained is about 0.23 eV, which compares well with the value of 0.25 eV (24.2 kJ/mol) [from Wikipedia](https://en.wikipedia.org/wiki/Pyramidal_inversion#Energy_barrier).

We might also notice the slight difference in energy between the two end-points of the path (about 1 meV). These should be identical, so any difference can be used as a very rough guide for the amount of numerical noise in the geometry optimisations.
If equivalent structures (the end-points in this example) were found to have very different energies, you would need to go back and tighten up your convergence criteria or force/energy tolerances in the initial geometry optimisations and re-run the NEB calculation with these tightened parameters. 

> Using our favourite visualisation software *[say what used here?]*, we can then look at how the structure evolves along the minimum energy pathway found via the NEB.

Reviewer: ```suggestion
Using our favourite visualisation software (here we used the [ASE POV-Ray interface](https://wiki.fysik.dtu.dk/ase/ase/io/io.html#examples)), we can then look at how the structure evolves along the minimum energy pathway found via the NEB.
```

 
The simplest way to do this is in the harmonic approximation. You need to estimate the harmonic zero-point energy at the start, transition and final configurations.

> For the particular case (ammonia) we have been looking at, it turns out that to actually explain the observed transition rate between the two 'umbrella' configurations, we need to go beyond the harmonic approximation ([see Wikipedia](https://en.wikipedia.org/wiki/Pyramidal_inversion#Quantum_effects)).  *[This is a bit obscure. Is it important for tutorial?]*

Reviewer: I wanted something at the end to tell the user what the next steps might be and to caution against taking the NEB barrier as gospel. Often the NEB barrier will need zero point energy corrections. These are most often done in the harmonic approximation. But even that wouldn't be enough to get an accurate barrier estimate for this NH3 example. 

Perhaps the ZPE correction instructions should be in a separate tutorial, but I wanted to flag it here as it's relevant for NEB barrier estimates.

Author: I get the point, but I think it's getting away from the concept of a tutorial on how to run NEB calculations towards a tutorial on calculating the energy barrier in NH3. I would simplify to a note that highlights that the NEB will provide a Delta E (this is exact), while the transition rates will depend on Delta E_0. The ZPE corrections may be significant  [presumably if the motion is dominated by H movement?]

Reviewer: ```suggestion
TODO: replace the section on zero-point energy corrections by a note that highlights that the NEB will provide a Delta E (this is exact), while the transition rates will depend on Delta E_0.
```
